beacon/light/canonical.go

```diff
@@ -0,0 +1,140 @@
+// with the keyPrefix in order to determine the ranges available in the database.
+func newCanonicalStore[T any](db ethdb.KeyValueStore, keyPrefix []byte,
+	encode func(T) ([]byte, error), decode func([]byte) (T, error)) *canonicalStore[T] {
+	cs := &canonicalStore[T]{
+		db:        db,
+		keyPrefix: keyPrefix,
+		encode:    encode,
+		decode:    decode,
+		cache:     lru.NewCache[uint64, T](100),
+	}
+	var (
+		iter = db.NewIterator(keyPrefix, nil)
+		kl   = len(keyPrefix)
+	)
+	for iter.Next() {
+		if len(iter.Key()) != kl+8 {
+			log.Warn("Invalid key length in the canonical chain database", "key", fmt.Sprintf("%#x", iter.Key()))
+			continue
+		}
+		period := binary.BigEndian.Uint64(iter.Key()[kl : kl+8])
+		if cs.periods.Start == 0 {
+			cs.periods.Start = period
+		} else if cs.periods.End != period {
+			log.Warn("Gap in the canonical chain database")
+			break // continuity guaranteed
+		}
+		cs.periods.End = period + 1
+	}
+	iter.Release()
+	return cs
+}
+
+// databaseKey returns the database key belonging to the given period.
+func (cs *canonicalStore[T]) databaseKey(period uint64) []byte {
+	var (
+		kl  = len(cs.keyPrefix)
+		key = make([]byte, kl+8)
+	)
+	copy(key[:kl], cs.keyPrefix)
+	binary.BigEndian.PutUint64(key[kl:], period)
+	return key
+}
+
+// add adds the given item to the database. It also ensures that the range remains
+// continuous. Can be used either with a batch or database backend.
+func (cs *canonicalStore[T]) add(backend ethdb.KeyValueWriter, period uint64, value T) error {
+	if !cs.periods.CanExpand(period) {
+		return fmt.Errorf("period expansion is not allowed, first: %d, next: %d, period: %d", cs.periods.Start, cs.periods.End, period)
+	}
+	enc, err := cs.encode(value)
+	if err != nil {
+		return err
+	}
+	if err := backend.Put(cs.databaseKey(period), enc); err != nil {
+		return err
+	}
+	cs.cache.Add(period, value)
+	cs.periods.Expand(period)
+	return nil
+}
+
+// deleteFrom removes items starting from the given period.
+func (cs *canonicalStore[T]) deleteFrom(batch ethdb.Batch, fromPeriod uint64) (deleted Range) {
+	if fromPeriod >= cs.periods.End {
+		return
+	}
+	if fromPeriod < cs.periods.Start {
+		fromPeriod = cs.periods.Start
+	}
+	deleted = Range{Start: fromPeriod, End: cs.periods.End}
+	for period := fromPeriod; period < cs.periods.End; period++ {
+		batch.Delete(cs.databaseKey(period))
+		cs.cache.Remove(period)
+	}
+	if fromPeriod > cs.periods.Start {
+		cs.periods.End = fromPeriod
+	} else {
+		cs.periods = Range{}
+	}
+	return
+}
+
+// get returns the item at the given period or the null value of the given type
+// if no item is present.
+// Note: get is thread safe in itself and therefore can be called either with
+// locked or unlocked chain mutex.
+func (cs *canonicalStore[T]) get(period uint64) (value T, ok bool) {
+	if value, ok = cs.cache.Get(period); ok {
+		return
+	}
+	if enc, err := cs.db.Get(cs.databaseKey(period)); err == nil {
```

Comment:
Is it intentional not to add the resolved value to the cache?

Reply:
It wasn't, fixed.